### ISSUE TITLE
study(rl): scalar energy reward 因果验证与统一评测

### ISSUE THREAD

**Issue by xcz0** [xcz0/Eco-AutoDrive#80]
## 背景

当前项目已经完成 PlannerRFT PPO-only 的机械链路、matched train/eval protocol 与第一轮 scalar reward 实验。

截至 #82 / E-031 / E-032，现有证据已经从“PPO 是否可训练”推进到一个新的问题：

- R0 在 `3 seeds × 100 updates` 下机械健康，但与 A0 held-out 行为基本不可区分；
- `λ ∈ {1,2,4,8}` 在 `1 seed × 50 updates` 下与 R0 及彼此基本不可区分；
- 标准 PPO 下长期呈现 `KL ~ 1e-8`、`clip_fraction=0`、ratio≈1、Beta≈初始化、guidance mean 仅 `1e-4~1e-3` 位移；
- 当前 no-traffic reward 中 TTC / Speed / safety gate 大量饱和，Progress 接近 1，Comfort 接近 0；
- 因此 #82 当前只能支持 `no measurable energy-term effect under the current PPO/reward regime`，不能支持“scalar energy objective 已失败”或“某个 λ 更优”。

另外，#81 已定位并修复 pre-existing PPO log-prob shape 广播缺陷：E-024~E-028 的训练发生在 cross-paired surrogate 下。修复后的 E-030~E-032 证明当前配置在标准 PPO 下稳定，但没有证明 E-028 `config-0001` 是“有效学习”的 optimizer 配置。

本 Issue 的研究目标因此调整为：

1. 证明 scalar objective 的变化可以被 PPO 区分并传递为不同 actor gradient；
2. 证明标准 PPO 不仅稳定，而且能产生非平凡的 learned behavioral effect；
3. 在此基础上再研究 energy weight `λ` 的 dose-response 与 energy / speed / progress / safety trade-off。

相关研究定义：

- `docs/research/README.md`
- `docs/research/reward-objective.md`
- `docs/research/ablation-plan.md`
- `docs/research/plannerrft-ppo/README.md`
- E-029：matched A0 frozen baseline
- E-030：标准 PPO 下 R0 / Rλ=1 transfer gate
- E-031：R0 3-seed ×100-update anchor
- E-032：`λ={1,2,4,8}` matched coarse sweep

## 核心 objective

继续使用：

```text
R0 = Gate × (5*TTC + 5*Progress + 2*Comfort + 4*Speed) / 16

Rλ = Gate × (5*TTC + 5*Progress + 2*Comfort + 4*Speed + λ*Energy) / (16 + λ)
```

最终比较仍为：

```text
A0 — Frozen Diffusion Planner
A1 — PPO + PlannerRFT R0
A2 — PPO + PlannerRFT Rλ
```

其中：

- `A1 vs A0`：判断 PPO + PlannerRFT objective 是否产生 learned behavioral effect；
- `A2 vs A1`：判断 energy term 是否提供独立增量收益。

## 控制变量原则

正式 λ 因果比较阶段原则上固定：

- frozen Diffusion Planner checkpoint / DDIM5 sampler；
- Exploration Policy 初始 checkpoint；
- observation / scene representation；
- training scenario family、seed namespace、transition 数；
- held-out evaluation scenarios / map seeds；
- evaluation horizon 与 termination semantics；
- reward-independent evaluation metrics。

但在新增的 #94 gate 中，允许**分阶段、单因素**调整：

1. reward normalization / component dynamic range；
2. 标准 PPO 的 effective-update 参数（重点 `lr / epochs / max_gradient_norm`）。

两者不得与 λ 主效应同时联合搜索。只有 #94 冻结新的 reward calibration 与 PPO control 后，#83 才恢复严格 matched λ comparison。

### Stage 1 — #81 实验协议、R0 与标准 PPO transfer gate【已完成主要交付】

- 建立 PlannerRFT no-energy `R0` reward profile；
- 固定 A0/A1/A2 matched train/eval protocol；
- 重跑 A0 frozen baseline（E-029）；
- 修复 PPO log-prob shape broadcasting；
- 在标准 PPO objective 下重跑 R0 / λ=1 transfer gate（E-030）。

结论：机械链路成立，可进入正式 reward study；不支持 behavioral improvement 结论。

### Stage 2 — #82 R0 anchor + 第一轮 λ coarse sweep【已完成，结论为 no measurable effect】

- R0：`3 seeds × 100 updates`（E-031）；
- `λ ∈ {1,2,4,8}`：各 `1 seed × 50 updates`（E-032）；
- matched held-out evaluation 与 policy diagnostics 完成。

当前结论：

```text
no measurable energy-term effect under current PPO/reward regime
```

本阶段**不再从 λ={1,2,4,8} 中强行选候选**，也不直接进入 #83。

### Stage 3 — #94 λ 可辨识性与 effective PPO update gate【当前优先级最高】

按顺序完成：

1. 同 rollout batch 的 `λ -> reward -> GAE -> normalized advantage -> actor gradient` identifiability 分析；
2. 如不可辨识，校准 Progress / Comfort / Energy 的 reward dynamic range；
3. 在标准 PPO objective 下搜索小型 effective-update region，重点考察 `lr / epochs / max_gradient_norm`；
4. 冻结一个“稳定且会学”的 PPO control；
5. 使用 `R0 vs λ=16` large-λ positive control 验证 objective 变化能够产生 guidance / speed / energy / progress 行为分离。

只有 #94 通过，才能进入 #83。

### Stage 4 — #83 正式 λ 候选筛选、多 seed confirmation 与最终 matched evaluation

#83 调整为依赖 #94，而不是直接依赖 #82 的 coarse sweep 候选。

进入 #83 后：

- 使用 #94 冻结的 reward calibration 与 effective PPO control；
- 重新进行正式 `λ={1,2,4,8}` dose-response / coarse sweep；
- 只有观察到可测、可解释的 λ-dependent behavior 后，才筛 1–2 个候选；
- 对候选做 `3 seeds × 100 updates`；
- A0 / A1 / A2 使用完全相同 held-out scenarios/seeds 做最终闭环 evaluation；
- 汇总 energy / progress / speed / safety / termination 与 matched differences。

## 统一评价指标

最终结论不得按 training return 排名。主表至少包含：

### Energy
- total proxy energy；
- energy intensity / normalized energy metric。

### Task / efficiency
- route progress；
- completion；
- distance；
- mean speed；
- stopped fraction；
- travel time / travel efficiency（若 artifact 支持）。

### Safety / legality
- collision；
- out-of-road；
- wrong-direction；
- termination type。

### PPO / behavior diagnostics
- longitudinal / lateral guidance distribution；
- Beta mean / concentration / variance；
- entropy；
- approximate KL；
- policy ratio / clip fraction；
- gradient diagnostics；
- Beta boundary / invalid action；
- episode length / reset frequency。

### λ identifiability（#94 新增）
- reward component variance；
- raw / normalized advantage correlation；
- advantage sign-flip fraction；
- actor gradient norm；
- gradient cosine similarity across λ。

## 结论判据

进入最终 λ comparison 前，必须先满足：

```text
λ changes normalized advantage / actor gradient
AND PPO produces non-trivial but stable policy updates
AND R0 vs large-λ stress arm produces measurable behavioral separation
```

最终可解释为 energy improvement 的基本形式：

```text
energy improves
AND safety does not materially regress
AND progress / completion remains comparable
AND speed / travel efficiency does not collapse
AND effect reproduces across matched seeds/scenarios
```

以下不计为成功：

- total energy 降低但 distance / progress 明显降低；
- 通过长时间停车或显著降速降低 energy；
- collision / out-of-road / failed termination 增加；
- 只有单一 seed 或个别场景改善；
- training reward 上升但独立 evaluation 无改善；
- λ 只改变 raw reward/value scale，却没有改变 normalized advantage / actor gradient / behavior。

## 完成判据

- [x] #81：R0、统一协议、A0 baseline、PPO shape 修复与标准 PPO transfer gate。
- [x] #82：R0 多 seed anchor 与第一轮 energy weight coarse sweep；结论为当前 regime 下 `no measurable effect`。
- [ ] #94：λ identifiability、reward dynamic-range gate、standard-PPO effective-update gate、R0 vs λ=16 positive control。
- [ ] #83：基于 #94 冻结配置重新做 λ dose-response、候选多 seed confirmation 与 A0/A1/A2 matched evaluation。
- [ ] 所有正式实验写入 `docs/experiments/records/`，记录 commit、dirty diff、resolved config、seeds、产物和结论边界。
- [ ] 最终能够明确回答 `A1 vs A0` 与 `A2 vs A1`，即使答案是“无显著改善”或“存在不可接受 trade-off”。

## 非目标

本 Issue 暂不进入：

- long-range preview / information ablation；
- encoder fine-tuning；
- multi-head critic / PPO-Lagrangian；
- GRPO / DiT direct fine-tuning；
- FASTSim 或真实动力系统能耗建模；
- 与 λ 联合的大规模 PPO hyperparameter search。

#94 中允许为建立 effective-update gate 做受控的小型 PPO 参数搜索，但不把 optimizer search 与最终 reward 主效应混在同一实验中。

**Comment by xcz0** [comment]
## #98 进展：Task A/B/C 完成

- Task A（E-043）：execution-horizon 干预支持归因 A（guidance temporal response × receding-horizon execution mismatch）。
- Task B（E-044）：learned guidance 方向误差主要由 longitudinal 分量贡献，横向分量很小，不支持横纵耦合为主因。
- Task C（E-045）：baseline 0.1 s receding horizon 下连续 20 个 replanning cycle 的逐 waypoint matched effect 显示稳定 `repeated_deferral`（13/16 场景）：正 response 的 zero-crossing 稳定在 0.2 s，越出执行的 0.1 s prefix，因此 planner full-horizon 正响应无法转化为 executed response。
- 依赖状态：#98 仍 **OPEN**，Task D（training adequacy / critic 离线归因）与 Task E（training-budget sweep）未执行；E-040 的 c4 failure 最终归因尚未完成，#83 继续 blocked。

**Comment by xcz0** [comment]
## #98 依赖同步

#98 已完成并关闭。已将 Stage 4 标记为 completed，并把 #83 的后续门槛从“等待 #98 归因”更新为“先在 #80 冻结正式 train/eval execution protocol”。E-047 已确认 `train–evaluation execution-contract mismatch mediated by guidance temporal response`，因此后续不再把 E-040 c4 failure 视为未解释问题。

**Comment by xcz0** [comment]
## 2026-09-24 Stage 5 split

为降低单 Issue 核验复杂度，Stage 5 已拆为：

#105 → #106 → #83

分别对应 canonical cadence PPO re-freeze、transfer/positive-control/λ dose-response、candidate multi-seed + final matched evaluation。父 Issue 正文已同步各自 gate、handoff 与 stop-path 逻辑。
